kernel/indTyping.ml

```diff
@@ -65,29 +65,20 @@ let mind_check_names mie =
 (************************************************************************)
 
 type univ_info = { ind_squashed : bool; ind_has_relevant_arg : bool;
-                   ind_min_univ : Sorts.t option; (* Some for template *)
+                   ind_template : bool;
                    ind_univ : Sorts.t;
                    missing : Sorts.t list; (* missing u <= ind_univ constraints *)
                  }
 
-let sup_sort s1 s2 = match s1, s2 with
-| (_, SProp) -> assert false (* template SProp not allowed *)
-| (SProp, s) | (Prop, s) | (s, Prop) -> s
-| (Set, Set) -> Sorts.set
-| (Set, Type u) | (Type u, Set) -> Sorts.sort_of_univ (Universe.sup u Universe.type0)
-| (Type u, Type v) -> Sorts.sort_of_univ (Universe.sup u v)
-
 let check_univ_leq ?(is_real_arg=false) env u info =
   let ind_univ = info.ind_univ in
   let info = if not info.ind_has_relevant_arg && is_real_arg && not (Sorts.is_sprop u)
     then {info with ind_has_relevant_arg=true}
     else info
   in
   (* Inductive types provide explicit lifting from SProp to other universes, so allow SProp <= any. *)
-  if Sorts.is_sprop u || UGraph.check_leq_sort (universes env) u ind_univ
-  then { info with ind_min_univ = Option.map (sup_sort u) info.ind_min_univ }
-  else if is_impredicative_sort env ind_univ
-       && Option.is_empty info.ind_min_univ then { info with ind_squashed = true }
+  if Sorts.is_sprop u || UGraph.check_leq_sort (universes env) u ind_univ then info
+  else if is_impredicative_sort env ind_univ then { info with ind_squashed = true }
```

Comment:
Shouldn't we still have `&& not info.ind_template` here? If we send the kernel something like
~~~coq
#[universes(template)]
Inductive strange (A:Type) : Prop := bla (_:A).
~~~
it probably shouldn't be accepted (although I don't know if it really matters)

Reply:
It's the other way around, making something squashed makes it less powerful because we don't allow elimination into Type. If anything this patch is making the kernel safer.

Reply:
(The next PR in line makes the kernel implementation of template poly much more similar to cumulpoly.)

Reply:
the else branch leads to an error so I don't see how it's safer, but if you're confident that it's safe enough then I'll accept it

Reply:
I am confident that template poly is currently unsound (although for unrelated reasons).